sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentUtils.scala

```diff
@@ -0,0 +1,204 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference, CreateNamedStruct, Expression, GetArrayItem, GetArrayStructFields, GetMapValue, GetStructField, Literal, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.logical.Assignment
+import org.apache.spark.sql.catalyst.util.CharVarcharUtils
+import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.types.{DataType, StructType}
+
+object AssignmentUtils extends SQLConfHelper with CastSupport {
+
+  private case class ColumnUpdate(ref: Seq[String], expr: Expression)
+
+  /**
+   * Aligns assignments to match table columns.
+   * <p>
+   * This method processes and reorders given assignments so that each target column gets
+   * an expression it should be set to. If a column does not have a matching assignment,
+   * it will be set to its current value. For example, if one passes table attributes c1, c2
+   * and an assignment c2 = 1, this method will return c1 = c1, c2 = 1. This alignment is
+   * required to construct an updated version of a row.
+   * <p>
+   * This method also handles updates to nested columns. If there is an assignment to a particular
+   * nested field, this method will construct a new struct with one field updated preserving other
+   * fields that have not been modified. For example, if one passes table attributes c1, c2
+   * where c2 is a struct with fields n1 and n2 and an assignment c2.n2 = 1, this method will
+   * return c1 = c1, c2 = struct(c2.n1, 1).
+   *
+   * @param attrs table attributes
+   * @param assignments assignments to align
+   * @return aligned assignments that match table attributes
+   */
+  def alignAssignments(
+      attrs: Seq[Attribute],
+      assignments: Seq[Assignment]): Seq[Assignment] = {
+
+    val errors = new mutable.ArrayBuffer[String]()
+
+    val output = applyUpdates(
+      updates = assignments.map(toColumnUpdate),
+      cols = attrs.map(restoreActualType),
+      colExprs = attrs,
+      addError = err => errors += err)
+
+    if (errors.nonEmpty) {
+      throw QueryCompilationErrors.invalidRowLevelOperationAssignments(assignments, errors.toSeq)
+    }
+
+    attrs.zip(output).map { case (attr, expr) => Assignment(attr, expr) }
+  }
+
+  private def toColumnUpdate(assignment: Assignment): ColumnUpdate = {
+    ColumnUpdate(toRef(assignment.key), assignment.value)
+  }
+
+  private def restoreActualType(attr: Attribute): Attribute = {
+    attr.withDataType(CharVarcharUtils.getRawType(attr.metadata).getOrElse(attr.dataType))
+  }
+
+  private def applyUpdates(
```

Comment:
> it's better to operate on the resolved column expressions

I agree, let's if we can avoid the conversion to references.

> We validate the map we got in step 2: for each top-level column, its expressions must be of the same tree height (to avoid updating both 'a.b' and 'a.b.c')

Could you elaborate a bit on how you see the tree heigh check? Like add a separate method for computing expression height? What about cases when it is OK to have different expression heights like 'a.b.n1' and 'a.c' where a, b, c are all structs?

Reply:
We could probably build `ExpressionSet` for each update key per top-level attribute and check the intersection across all `ExpressionSet` is empty. Let me know if that's similar to what you thought.

Reply:
@cloud-fan, any ideas on how to avoid deconstructing `Seq[String]` when applying a set of assignments to a top-level attribute? The problem there is that we recurse top to bottom in `applyUpdates` whereas `assignment.key` is a set of nested GetStructField calls with the outer expression referring the leaf column.

Reply:
I'm thinking about this
```
def alignAssignments(
    assignments: Seq[Assignment],
    attrs: Seq[Attribute]): Seq[Assignment] = {
  // use ExpressionSet to check assignments have no duplication
  ...
  attrs.map { attr =>
    Assignment(attr, applyUpdates(assignments, attr))
  }
}

def applyUpdates(
    assignments: Seq[Assignment],
    col: Expression): Expression = {
  val (exactAssigments, others) = assignments.partition { assignment =>
    assigment.key.semanticEquals(col)
  }
  val relatedAssignments = others.filter { assignment =>
    assigment.key.exists(_.semanticEquals(col))
  }
  assert(exactAssigments.length <= 1)
  if (exactAssigments.nonEmpty) {
    if (relatedAssignments.nonEmpty) fail...
    exactAssigments.head.value
  } else {
    col.dataType match {
      case StructType(fields) =>
        CreatedStruct(fields.flatMap { field =>
          applyUpdates(relatedAssignments, GetStruct(col, field.name))
        })

      case _ =>
        assert(relatedAssignments.isEmpty)
        col
    }
  }
}
```

Reply:
Perfect, I forgot about `exists`. Thanks!